mypy/typeanal.py

```diff
@@ -363,36 +364,85 @@ def analyze_callable_type(self, t: UnboundType) -> Type:
         elif len(t.args) == 2:
-            ret_type = self.anal_type(t.args[1])
+            ret_type = t.args[1]
             if isinstance(t.args[0], TypeList):
                 # Callable[[ARG, ...], RET] (ordinary callable type)
                 args = t.args[0].items
-                return CallableType(self.anal_array(args),
-                                    [nodes.ARG_POS] * len(args),
-                                    [None] * len(args),
-                                    ret_type=ret_type,
-                                    fallback=fallback)
+                ret = CallableType(args,
+                                   [nodes.ARG_POS] * len(args),
+                                   [None] * len(args),
+                                   ret_type=ret_type,
+                                   fallback=fallback)
             elif isinstance(t.args[0], EllipsisType):
                 # Callable[..., RET] (with literal ellipsis; accept arbitrary arguments)
-                return CallableType([AnyType(), AnyType()],
-                                    [nodes.ARG_STAR, nodes.ARG_STAR2],
-                                    [None, None],
-                                    ret_type=ret_type,
-                                    fallback=fallback,
-                                    is_ellipsis_args=True)
+                ret = CallableType([AnyType(), AnyType()],
+                                   [nodes.ARG_STAR, nodes.ARG_STAR2],
+                                   [None, None],
+                                   ret_type=ret_type,
+                                   fallback=fallback,
+                                   is_ellipsis_args=True)
             else:
                 self.fail('The first argument to Callable must be a list of types or "..."', t)
                 return AnyType()
-
-        self.fail('Invalid function type', t)
-        return AnyType()
+        else:
+            self.fail('Invalid function type', t)
+            return AnyType()
+        assert isinstance(ret, CallableType)
+        return ret.accept(self)
+
+    @contextmanager
+    def tvar_scope_frame(self) -> Generator:
+        old_scope = self.tvar_scope
+        self.tvar_scope = TypeVarScope(self.tvar_scope)
+        yield
+        self.tvar_scope = old_scope
+
+    def infer_type_variables(self,
+                             type: CallableType) -> List[Tuple[str, TypeVarExpr]]:
+        """Return list of unique type variables referred to in a callable."""
+        names = []  # type: List[str]
+        tvars = []  # type: List[TypeVarExpr]
+        for arg in type.arg_types:
+            for name, tvar_expr in arg.accept(TypeVariableQuery(self.lookup, self.tvar_scope)):
+                if name not in names:
+                    names.append(name)
+                    tvars.append(tvar_expr)
+        for name, tvar_expr in type.ret_type.accept(
+                TypeVariableQuery(self.lookup, self.tvar_scope, include_callables=False)):
```

Comment:
Other punchline of this diff: when you're finding the type variables that a function has, don't include the tvars you *only* find in a `Callable` in the return type.  Those belong to that `Callable`